# Connecticut Real Estate 2011-2021
### Aaron 'Asa' Sherwyn
# 02 Wrangling
1. Imports and Set-up
1. Pandemic Flag
1. Deriving Variables
    - Reducing date to yyyy and mm only
    - Reducing address details street_number and street_suffix only
1. Merging Data Sets
    - Town data
    - Consumer Price Index
1. Inflation Normalizations
    - Normalized assessed_value
    - Normalized sale_amount
    - Normalized sales_ratio
4. Export

## 1. Imports and Set-up

In [1]:
#loading libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Creating a path to the root directory for this project
path = r'C:\Users\theva\OneDrive\CareerFoundry\Data Immersion\A6\Connecticut Real Estate 2001-2020'

In [3]:
#Importing main dataframe
df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'CTRE 2011-2021 CLEANED.pkl'))

In [4]:
#Importing additional town data
df_town = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'CT Town Data.csv'))

In [5]:
#Importing Consumer Price Index
df_cpi = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'CPI 2010-2021.csv'))

In [6]:
#Increasing display limit
pd.set_option('display.max_rows', 200)

In [7]:
#Checking df
df.head(20)

serial_number  list_year date_recorded        town                address  \
0         2020348       2020    2021-09-13     Ansonia        230 WAKELEE AVE   
1           20002       2020    2020-10-02     Ashford        390 TURNPIKE RD   
2          200212       2020    2021-03-09        Avon       5 CHESTNUT DRIVE   
3          200243       2020    2021-04-13        Avon  111 NORTHINGTON DRIVE   
4          200377       2020    2021-07-02        Avon     70 FAR HILLS DRIVE   
5          200109       2020    2020-12-09        Avon     57 FAR HILLS DRIVE   
7         2020313       2020    2021-07-01      Berlin      216 WATCH HILL RD   
8          200097       2020    2021-06-04     Bethany            23 AMITY RD   
9           20139       2020    2020-12-16      Bethel      16 DEEPWOOD DRIVE   
10         200086       2020    2021-08-10   Bethlehem         39 WOODLAND RD   
11        2000381       2020    2021-09-13  Bloomfield         9 SADDLE RIDGE   
12         200258       2020    2021-01-22    Branford   34 SUNRISE COVE CAMP   
13         200804       2020    2021-09-28    Branford       119 MONTOWESE ST   
14         201295       2020    2021-09-09     Bristol            609 CAMP ST   
15         200032       2020    2020-10-08     Bristol         9 LESLIE COURT   
16         200354       2020    2020-12-29     Bristol         391 TIFFANY LA   
17         200527       2020    2021-03-01     Bristol             25 WARD ST   
18         201212       2020    2021-08-23     Bristol        8 JENNINGS TERR   
19       20000050       2020    2020-11-05  Brookfield           4 CHELSEA CT   
20       20000179       2020    2021-02-22  Brookfield        72 HOMESTEAD LN   

    assessed_value  sale_amount  sales_ratio property_type residential_type  
0           150500     325000.0     0.463000    Commercial              NaN  
1           253000     430000.0     0.588300   Residential    Single Family  
2           130400     179900.0     0.724800   Residential            Condo  
3           619290     890000.0     0.695800   Residential    Single Family  
4           862330    1447500.0     0.595700   Residential    Single Family  
5           847520    1250000.0     0.678000   Residential    Single Family  
7           412000     677500.0     0.608100   Residential    Single Family  
8           511000     795000.0     0.642700    Commercial              NaN  
9           171360     335000.0     0.511500   Residential    Single Family  
10          168900     352000.0     0.479800   Residential    Single Family  
11          163730     250000.0     0.654900   Residential            Condo  
12          530500     700000.0     0.757800   Residential    Single Family  
13          455100     650000.0     0.700100    Commercial              NaN  
14          144340     249900.0     0.577500   Residential    Single Family  
15           94920     155000.0     0.612300   Residential    Single Family  
16          173740     299000.0     0.581000   Residential    Single Family  
17           85680     183000.0     0.468100   Residential       Two Family  
18           88060      91000.0     0.967600   Residential    Single Family  
19          105110     160000.0     0.656937   Residential            Condo  
20           94770     170000.0     0.557400   Residential            Condo

In [8]:
#Checking df
df.tail(20)

serial_number  list_year date_recorded           town  \
508377         100012       2010    2010-10-25     Woodbridge   
508379          10155       2010    2010-12-20        Milford   
508381         100027       2010    2010-11-01       Branford   
508382         100117       2010    2010-10-15      Newington   
508383          10146       2010    2011-06-21     Bloomfield   
508384          10212       2010    2011-03-16         Hamden   
508386          10055       2010    2010-12-22    North Haven   
508387         100846       2010    2011-06-28       Stamford   
508388          10088       2010    2010-12-20         Groton   
508389          10076       2010    2011-03-04     Brookfield   
508390         100077       2010    2011-03-07   Old Saybrook   
508391          10019       2010    2010-10-29      Watertown   
508392         100170       2010    2010-12-20  West Hartford   
508395         100358       2010    2011-04-29      Stratford   
508396         100169       2010    2011-06-29     Plainville   
508397         100125       2010    2011-01-10        Norwich   
508398          10033       2010    2010-11-05  South Windsor   
508399         100027       2010    2010-12-20    East Granby   
508401         100266       2010    2011-07-18     Farmington   
508404         100252       2010    2011-03-31        Meriden   

                        address  assessed_value  sale_amount  sales_ratio  \
508377         34 WEPAWAUG ROAD          284900     415000.0     0.686506   
508379      4 LOOKOUT HILL ROAD          269100     310000.0     0.868065   
508381      346 JEFFERSON WOODS          111100     139000.0     0.799281   
508382       85 WOODSEDGE DRIVE           65310      97000.0     0.673299   
508383           19 CLAIRE LANE          358470     560000.0     0.640125   
508384        4 SKY VIEW CIRCLE          181160     207500.0     0.873060   
508386           9 ROBERTA ROAD          168980     190000.0     0.889368   
508387    143 PALMERS HILL ROAD          738830     687500.0     1.074662   
508388              1 JOLIET CT          114240     128000.0     0.892500   
508389  89A LONG MEADOW HILL RD          349480     439000.0     0.796082   
508390       191 INGHAM HILL RD          318000     340000.0     0.935294   
508391            65 EDWARD AVE          161600     170000.0     0.950588   
508392    534 SOUTH QUAKER LANE          122700     187000.0     0.656150   
508395          190 NICHOLS AVE          160510     221000.0     0.726290   
508396     190 TOMLINSON AVE 3C          131740     149000.0     0.884161   
508397            246 ROGERS RD          305000     350000.0     0.871429   
508398             142 SCOTT DR          242690     325000.0     0.746738   
508399  16 SEYMOUR RD., UNIT 5F           81900     113000.0     0.724779   
508401           45 GARDEN GATE          537330     730000.0     0.736068   
508404               40 LYNN DR          148070     188000.0     0.787606   

       property_type residential_type  
508377   Residential    Single Family  
508379   Residential            Condo  
508381   Residential            Condo  
508382   Residential            Condo  
508383   Residential    Single Family  
508384   Residential    Single Family  
508386   Residential    Single Family  
508387   Residential    Single Family  
508388   Residential    Single Family  
508389   Residential    Single Family  
508390   Residential    Single Family  
508391   Residential    Single Family  
508392   Residential    Single Family  
508395   Residential    Single Family  
508396   Residential            Condo  
508397   Residential    Single Family  
508398   Residential    Single Family  
508399   Residential            Condo  
508401   Residential    Single Family  
508404   Residential    Single Family

In [9]:
#Checking df
df.shape

(435179, 10)

In [10]:
#Checking df_town
df_town.head(20)

town  population  latitude  longitude
0        Andover        2998   41.7373   -72.3704
1        Ansonia       18945   41.3462   -73.0790
2        Ashford        4193   41.8731   -72.1216
3           Avon       18821   41.8096   -72.8305
4    Barkhamsted        3659   41.9293   -72.9140
5   Beacon Falls        6027   41.4426   -73.0618
6         Berlin       20107   41.6215   -72.7457
7        Bethany        5331   41.4218   -72.9970
8         Bethel       20287   41.3712   -73.4140
9      Bethlehem        3408   41.6404   -73.2058
10    Bloomfield       21399   41.8265   -72.7301
11        Bolton        4839   41.7690   -72.4334
12        Bozrah        2389   41.5604   -72.1636
13      Branford       28230   41.2799   -72.8141
14    Bridgeport      148529   41.1792   -73.1894
15   Bridgewater        1824   41.5351   -73.3662
16       Bristol       60733   41.6718   -72.9493
17    Brookfield       17428   41.4826   -73.4096
18      Brooklyn        8421   41.7881   -71.9497
19    Burlington        9511   41.7691   -72.9648

In [11]:
#Checking df_town
df_town.tail(20)

town  population  latitude  longitude
149     Washington        3619   41.6303   -73.3107
150      Waterbury      113783   41.5582   -73.0515
151      Waterford       19558   41.3542   -72.1648
152      Watertown       22110   41.6062   -73.1183
153  West Hartford       64034   41.7621   -72.7420
154     West Haven       55518   41.2705   -72.9470
155      Westbrook        6800   41.2960   -72.4642
156         Weston       10339   41.2009   -73.3807
157       Westport       27168   41.1415   -73.3579
158   Wethersfield       27162   41.7064   -72.6596
159     Willington        5586   41.8809   -72.2601
160         Wilton       18486   41.1954   -73.4379
161     Winchester       10335   41.8994   -73.1348
162        Windham       24480   41.6998   -72.1570
163        Windsor       29424   41.8526   -72.6437
164  Windsor Locks       12559   41.9243   -72.6454
165        Wolcott       16232   41.6023   -72.9868
166     Woodbridge        9073   41.3526   -73.0084
167       Woodbury        9762   41.5445   -73.2090
168      Woodstock        8187   41.9484   -71.9740

In [12]:
#Checking df_town
df_town.shape

(169, 4)

In [13]:
#Checking df_town
df_town.columns

Index(['town', 'population', 'latitude', 'longitude'], dtype='object')

In [14]:
#Checking df_cpi
df_cpi.head(20)

date  price_index
0   2010-01      212.568
1   2010-02      212.544
2   2010-03      213.525
3   2010-04      213.958
4   2010-05      214.124
5   2010-06      213.839
6   2010-07      213.898
7   2010-08      214.205
8   2010-09      214.306
9   2010-10      214.623
10  2010-11      214.750
11  2010-12      215.262
12  2011-01      216.400
13  2011-02      217.535
14  2011-03      220.024
15  2011-04      221.743
16  2011-05      222.954
17  2011-06      222.522
18  2011-07      222.686
19  2011-08      223.326

In [15]:
#Checking df_cpi
df_cpi.tail(20)

date  price_index
124  2020-05      249.521
125  2020-06      251.054
126  2020-07      252.636
127  2020-08      253.597
128  2020-09      254.004
129  2020-10      254.076
130  2020-11      253.826
131  2020-12      254.081
132  2021-01      255.296
133  2021-02      256.843
134  2021-03      258.935
135  2021-04      261.237
136  2021-05      263.612
137  2021-06      266.412
138  2021-07      267.789
139  2021-08      268.387
140  2021-09      269.086
141  2021-10      271.552
142  2021-11      273.042
143  2021-12      273.925

In [16]:
#Checking df_cpi
df_cpi.shape

(144, 2)

In [17]:
#Checking df_cpi
df_cpi.columns

Index(['date', 'price_index'], dtype='object')

## 2. Pandemic Flag

In [18]:
#creating in_pandemic variable

p_date = pd.Timestamp(2020, 3, 11,0) #Official start of the pandemic is March 3, 2020
df['in_pandemic'] = 0 #Making default in_pandemic value 0
df.loc[df['date_recorded'] >= p_date, 'in_pandemic'] = 1 #if date_recorded is >= p_date, in_pandemic is 1

In [19]:
#Checking results

df[['date_recorded', 'in_pandemic']].sample(40)

date_recorded  in_pandemic
323030    2015-03-30            0
154667    2019-07-17            0
348590    2014-10-01            0
33498     2020-10-19            1
387874    2014-07-28            0
211889    2017-11-21            0
87564     2020-09-18            1
437017    2012-11-05            0
76857     2020-01-13            0
412930    2013-06-25            0
332444    2015-06-09            0
273233    2016-03-31            0
784       2021-04-05            1
139997    2019-02-27            0
77945     2020-08-25            1
188236    2018-09-10            0
54463     2020-12-22            1
80333     2020-06-30            1
311319    2016-06-09            0
71017     2020-08-31            1
312017    2016-05-31            0
369331    2013-10-04            0
111976    2020-06-15            1
264623    2017-04-17            0
142554    2019-04-01            0
332480    2015-07-30            0
262409    2017-05-23            0
356092    2015-06-12            0
372648    2014-05-01            0
486535    2011-05-31            0
36339     2020-10-13            1
226925    2017-09-26            0
305253    2016-06-17            0
28452     2021-09-01            1
87007     2019-12-12            0
271716    2017-09-29            0
65115     2021-07-15            1
139649    2018-10-22            0
379859    2014-03-04            0
424477    2013-06-04            0

In [20]:
#Checking results

df['in_pandemic'].value_counts()

0    339061
1     96118
Name: in_pandemic, dtype: int64

## 3. Deriving Variables

### Reducing date to yyyy and mm only

In [21]:
#Creating date variable with just year and month to help protect privacy
df['month'] = df['date_recorded'].dt.strftime('%m').astype(int)

In [22]:
#Creating date variable with just year and month to help protect privacy
df['year'] = df['date_recorded'].dt.strftime('%Y').astype(int)

In [23]:
#checking new variable
df[['date_recorded', 'month', 'year']].head(20)

date_recorded  month  year
0     2021-09-13      9  2021
1     2020-10-02     10  2020
2     2021-03-09      3  2021
3     2021-04-13      4  2021
4     2021-07-02      7  2021
5     2020-12-09     12  2020
7     2021-07-01      7  2021
8     2021-06-04      6  2021
9     2020-12-16     12  2020
10    2021-08-10      8  2021
11    2021-09-13      9  2021
12    2021-01-22      1  2021
13    2021-09-28      9  2021
14    2021-09-09      9  2021
15    2020-10-08     10  2020
16    2020-12-29     12  2020
17    2021-03-01      3  2021
18    2021-08-23      8  2021
19    2020-11-05     11  2020
20    2021-02-22      2  2021

In [24]:
#Checking new variables
df[['month', 'year']].describe()

month           year
count  435179.000000  435179.000000
mean        6.881538    2016.586894
std         3.195758       3.129705
min         1.000000    1999.000000
25%         4.000000    2014.000000
50%         7.000000    2017.000000
75%         9.000000    2019.000000
max        12.000000    2021.000000

In [25]:
#checking new variable frequencies
df['month'].value_counts().sort_index()

1     26159
2     22742
3     29186
4     32029
5     38285
6     45866
7     47127
8     46709
9     39276
10    38066
11    33750
12    35984
Name: month, dtype: int64

In [26]:
#checking new variable frequencies
df['year'].value_counts().sort_index()

1999        1
2001        2
2008        4
2010     6559
2011    24697
2012    25432
2013    30910
2014    34417
2015    40706
2016    40222
2017    41748
2018    40822
2019    46129
2020    57109
2021    46421
Name: year, dtype: int64

>Notes: year should not have 1999, 2001, or 2008...investigating

In [27]:
#looking at anomalous years more closely
df.loc[df['year'] < 2010]

serial_number  list_year date_recorded           town  \
178779         170164       2017    2008-05-29        Clinton   
183429          17698       2017    2008-08-02  East Hartford   
185557         170107       2017    2008-06-13      Westbrook   
192062         170177       2017    1999-04-05     New London   
198408         170207       2017    2008-07-03        Clinton   
211498         173165       2017    2001-08-01        Shelton   
213100         173202       2017    2001-08-23        Shelton   

                   address  assessed_value  sale_amount  sales_ratio  \
178779        6 OAKWOOD LN          116700     187500.0     0.622400   
183429    1073 TOLLAND ST.          101230     157000.0     0.644777   
185557      84 PLYMOUTH RD          192500     304000.0     0.633224   
192062  184 PEQUOT AVE 309           46690      95000.0     0.491400   
198408   33 CARTER HILL RD          220400     330000.0     0.667879   
211498    19 BLACKBERRY LA          390810     635000.0     0.615449   
213100   12 BOYSENBERRY LA          427910     558000.0     0.766864   

       property_type residential_type  in_pandemic  month  year  
178779   Residential    Single Family            0      5  2008  
183429   Residential    Single Family            0      8  2008  
185557   Residential    Single Family            0      6  2008  
192062   Residential            Condo            0      4  1999  
198408   Residential    Single Family            0      7  2008  
211498   Residential    Single Family            0      8  2001  
213100   Residential    Single Family            0      8  2001

In [28]:
#All unusual years are from list_year 2017, thus for their given months their year should be 2018
#changing years < 2010 to year = 2018
df.loc[df['year'] < 2010, 'year'] = 2018

In [29]:
#re-checking new variable frequencies
df['year'].value_counts().sort_index()

2010     6559
2011    24697
2012    25432
2013    30910
2014    34417
2015    40706
2016    40222
2017    41748
2018    40829
2019    46129
2020    57109
2021    46421
Name: year, dtype: int64

In [30]:
#checking data types
df.dtypes

serial_number                int64
list_year                    int64
date_recorded       datetime64[ns]
town                        object
address                     object
assessed_value               int64
sale_amount                float64
sales_ratio                float64
property_type               object
residential_type            object
in_pandemic                  int64
month                        int32
year                         int32
dtype: object

In [31]:
#dropping date_recorded from the dataframe
df.drop(columns = ['date_recorded'], inplace = True)

In [32]:
#checking current shape
df.shape

(435179, 12)

### Reducing address details street_number and street_suffix only

In [33]:
#creating a variable for the street number
df['street_number'] = df['address'].str.split().str.get(0)

In [34]:
#converting street_number to numeric values
df['street_number'] = pd.to_numeric(df['street_number'], errors = 'coerce')

In [35]:
#checking frequencies
df['street_number'].value_counts(dropna = False)

NaN         15565
15.0         6164
5.0          5809
11.0         5774
10.0         5654
            ...  
5969.0          1
151900.0        1
3321.0          1
4611.0          1
4022.0          1
Name: street_number, Length: 3410, dtype: int64

In [36]:
#examining street numbers > 10,000
df.loc[df['street_number'] > 10000]

serial_number  list_year           town  \
69             200579       2020      Greenwich   
924            201699       2020      Greenwich   
6263           200260       2020   East Windsor   
7152           200630       2020      Greenwich   
9194           200530       2020      Greenwich   
15069          201300       2020      Greenwich   
16465          200477       2020      Greenwich   
18563          201780       2020      Greenwich   
30374          200049       2020     Middlebury   
33068          200063       2020        Meriden   
57906          200397       2020       Trumbull   
58916          200249       2020       Trumbull   
64177          200671       2020       Trumbull   
65362          200032       2020       Trumbull   
66182          200585       2020       Trumbull   
80750          190204       2019      Killingly   
81056          190351       2019      Greenwich   
101711         190530       2019       Trumbull   
104227         190270       2019     New Canaan   
139721          18335       2018         Monroe   
168275         180510       2018       Trumbull   
168639          18289       2018     West Haven   
168771         180435       2018       Trumbull   
188268         170366       2017       Trumbull   
213311         170282       2017     Torrington   
218784         170028       2017       Trumbull   
220805         170320       2017       Trumbull   
228239        1600093       2016       Guilford   
239825         161212       2016       Woodbury   
244142         160225       2016       Trumbull   
247821          16404       2016  East Hartford   
255795         160014       2016         Putnam   
259481         160548       2016       Trumbull   
263141         160217       2016       Trumbull   
267919         160300       2016       Trumbull   
268773         161243       2016      Waterbury   
281570         150134       2015   East Windsor   
283688          15031       2015        Milford   
333021          14777       2014        Milford   
353948         140059       2014       Woodbury   
359180         140014       2014       Trumbull   
365329         140173       2014       Plymouth   
376493         131567       2013     Bridgeport   
390272          13386       2013     West Haven   
407083         130056       2013       Trumbull   
407085         130546       2013     Torrington   
417435         120447       2012     Middletown   
419742         120598       2012     Bridgeport   
435406         121900       2012      Stratford   
471776         110067       2011       Suffield   
499028         100298       2010       Trumbull   

                              address  assessed_value  sale_amount  \
69                            2195000         1061900    2195000.0   
924                            720000          536200     720000.0   
6263                            88510          268100     497500.0   
7152                          2350000         1964480    2350000.0   
9194                          1050000          521500    1050000.0   
15069                          920000          515900     920000.0   
16465                         4065110         1675170    4910000.0   
18563                          980210         1400300    1650000.0   
30374                          332900          218200     332900.0   
33068                          143000           76790     143000.0   
57906               22122 ARGANESE PL          139440     225000.0   
58916            12214 ARGANESE PLACE          127540     215000.0   
64177            12215 ARGANESE PLACE          128730     235000.0   
65362            11210 ARGANESE PLACE          128730     215000.0   
66182            11209 ARGANESE PLACE          133980     219000.0   
80750                          152500           91910     152500.0   
81056                          715000          540320     715000.0   
101711           11212 ARGANESE PLACE          128240     204000.0   
104227                   

In [37]:
#examining street_numbers > 7,500 and < 10,000
df.loc[(df['street_number'] > 7500) & (df['street_number'] < 10000)]

serial_number  list_year           town                      address  \
63              20054       2020     Farmington                  8193 ELY RD   
2260            20023       2020     Farmington           9391 NORTHWOODS RD   
2304           200480       2020     Farmington            8531 MT SPRING RD   
2897           200664       2020     Farmington             8649 PINNACLE RD   
5383           200554       2020     Farmington                 8451 MAIN ST   
5904           200687       2020     Farmington          8084 CEDAR RIDGE DR   
6288           200286       2020     Farmington  9278 BRIDGEHAMPTON CROSSING   
6322           200173       2020     Farmington               8027 BEN COURT   
6417           200529       2020     Farmington          8556 NEW BRITAIN AV   
7039           200382       2020     Farmington           9325 GATESHEAD WAY   
7863           200182       2020     Farmington                8468 MAPLE AV   
8133           200145       2020     Farmington        8631 PHEASANT HILL RD   
9631            20099       2020     Farmington         9318 NEWCASTLE PLACE   
10898          200542       2020     Farmington               8090 CIRCLE DR   
10944           20098       2020     Farmington                8864 SOUTH RD   
11732          200446       2020     Farmington            9030 WEST GATE RD   
11886          200601       2020     Farmington     8252 FARMINGTON RIDGE DR   
13211          200695       2020     Farmington        8938 THATCHER TERRACE   
13789          200545       2020     Farmington           9014 WATERVILLE RD   
15067          200584       2020     Farmington  9275 BRIDGEHAMPTON CROSSING   
15328          200387       2020     Farmington                  8203 ELY RD   
16684          200340       2020     Farmington           9523 PLAINVILLE AV   
18194          200256       2020     Farmington                  8201 ELY RD   
18285          200546       2020     Farmington           9012 WATERVILLE RD   
18622          200249       2020     Farmington               8491 MEADOW RD   
20050          200179       2020     Farmington           8980 TWIN PONDS DR   
22087          200287       2020     Farmington  9275 BRIDGEHAMPTON CROSSING   
22609          200357       2020     Farmington                  8195 ELY RD   
22810          200303       2020     Farmington           9324 GATESHEAD WAY   
23264           20013       2020     Farmington            9048 WHITE OAK RD   
36778          200394       2020        Meriden         7913 HOLLOW BROOK II   
37806          200091       2020       Portland         9813 WANGUNK MEADOWS   
39354          200842       2020        Meriden         7932 HOLLOW BROOK II   
42261          200093       2020       Portland         9805 WANGUNK MEADOWS   
72653           19034       2019     Farmington               8410 JUNIOR RD   
100950         190546       2019    Southington             8811 GLACIER WAY   
114665         190462       2019        Meriden         7934 HOLLOW BROOK II   
118288         190554       2019        Meriden         7952 HOLLOW BROOK II   
140637          18189       2018     Farmington         8045 BLODGETT ROY DR   
142858         180715       2018        Meriden         7915 HOLLOW BROOK II   
161664         180522       2018        Meriden         7951 HOLLOW BROOK II   
174381         180738       2018    Southington             8810 GLACIER WAY   
176397         170079       2017        Meriden            7914 HOLLOW BROOK   
202574         170499       2017    Southington         9310 SWEETHEART PATH   
210556         170281       2017  West Hartford      7995 PROSPECT AVENUE A5   
213521         170108       2017    Southington             8811 GLACIER WAY   
225382         160659       2016        Meriden         7932 HOLLOW BROOK II   
232741         160559       2016        Meriden         7917 HOLLOW BROOK II   
266909          16034       2016         Orange            8360 BEECHWOOD RD   
278173  

In [38]:
#Making unusual entries NaN (street_number > 10000)
df.loc[df['street_number'] > 10000, 'street_number'] = np.nan

In [39]:
#checking frequencies
df['street_number'].value_counts(dropna = False)

NaN       15616
15.0       6164
5.0        5809
11.0       5774
10.0       5654
          ...  
6073.0        1
4319.0        1
3077.0        1
5312.0        1
4022.0        1
Name: street_number, Length: 3363, dtype: int64

In [40]:
#checking new variable
df[['address', 'street_number']].head(20)

address  street_number
0         230 WAKELEE AVE          230.0
1         390 TURNPIKE RD          390.0
2        5 CHESTNUT DRIVE            5.0
3   111 NORTHINGTON DRIVE          111.0
4      70 FAR HILLS DRIVE           70.0
5      57 FAR HILLS DRIVE           57.0
7       216 WATCH HILL RD          216.0
8             23 AMITY RD           23.0
9       16 DEEPWOOD DRIVE           16.0
10         39 WOODLAND RD           39.0
11         9 SADDLE RIDGE            9.0
12   34 SUNRISE COVE CAMP           34.0
13       119 MONTOWESE ST          119.0
14            609 CAMP ST          609.0
15         9 LESLIE COURT            9.0
16         391 TIFFANY LA          391.0
17             25 WARD ST           25.0
18        8 JENNINGS TERR            8.0
19           4 CHELSEA CT            4.0
20        72 HOMESTEAD LN           72.0

In [41]:
#checking current shape
df.shape

(435179, 13)

>Notes: 15,565 Entries had non-numeric street numbers and were converted to NaN in street_number. Street numbers larger than 10,000 had errors (were sales_amount, lacked full address, or did not correspond to an actual listed address via google maps search) were made NaN.

In [42]:
#creating a variable for the street suffix
df['street_suffix'] = df['address'].str.split().str.get(-1)

In [43]:
#checking new variable
df[['address', 'street_suffix']].head(20)

address street_suffix
0         230 WAKELEE AVE           AVE
1         390 TURNPIKE RD            RD
2        5 CHESTNUT DRIVE         DRIVE
3   111 NORTHINGTON DRIVE         DRIVE
4      70 FAR HILLS DRIVE         DRIVE
5      57 FAR HILLS DRIVE         DRIVE
7       216 WATCH HILL RD            RD
8             23 AMITY RD            RD
9       16 DEEPWOOD DRIVE         DRIVE
10         39 WOODLAND RD            RD
11         9 SADDLE RIDGE         RIDGE
12   34 SUNRISE COVE CAMP          CAMP
13       119 MONTOWESE ST            ST
14            609 CAMP ST            ST
15         9 LESLIE COURT         COURT
16         391 TIFFANY LA            LA
17             25 WARD ST            ST
18        8 JENNINGS TERR          TERR
19           4 CHELSEA CT            CT
20        72 HOMESTEAD LN            LN

In [44]:
#checking current shape
df.shape

(435179, 14)

In [46]:
#finding most popular street suffixes
df['street_suffix'].value_counts().loc[lambda x : x>1000]

RD         87915
ST         57494
DR         41028
AVE        33254
ROAD       30802
STREET     15557
DRIVE      15551
LANE       14322
LN         13373
AVENUE      8752
LA          6083
CT          5182
WAY         4423
COURT       3358
CIRCLE      3150
CIR         3093
PL          2879
PLACE       2363
AV          2314
VILLAGE     2200
TRAIL       1838
TERRACE     1640
BLVD        1388
EXT         1352
TERR        1347
RIDGE       1192
TER         1160
TPKE        1140
ST.         1139
HILL        1078
Name: street_suffix, dtype: int64

In [47]:
#unifying street suffix (with 1000+ counts) variations according to USPS standards
df.loc[(df['street_suffix'] == 'RD') |
       (df['street_suffix'] == 'RD.'), 'street_suffix'] = 'ROAD'
df.loc[(df['street_suffix'] == 'ST') |
       (df['street_suffix'] == 'ST.'), 'street_suffix'] = 'STREET'
df.loc[(df['street_suffix'] == 'DR') |
       (df['street_suffix'] == 'DR.'), 'street_suffix'] = 'DRIVE'
df.loc[(df['street_suffix'] == 'LA') |
       (df['street_suffix'] == 'LN'), 'street_suffix'] = 'LANE'
df.loc[(df['street_suffix'] == 'AVE') |
       (df['street_suffix'] == 'AV'), 'street_suffix'] = 'AVENUE'
df.loc[(df['street_suffix'] == 'CT'), 'street_suffix'] = 'COURT'
df.loc[(df['street_suffix'] == 'CIR') |
       (df['street_suffix'] == 'CIRC') |
       (df['street_suffix'] == 'CIRCL') |
       (df['street_suffix'] == 'CRCL') |
       (df['street_suffix'] == 'CRCLE'), 'street_suffix'] = 'CIRCLE'
df.loc[(df['street_suffix'] == 'PL'), 'street_suffix'] = 'PLACE'
df.loc[(df['street_suffix'] == 'VLG'), 'street_suffix'] = 'VILLAGE'
df.loc[(df['street_suffix'] == 'TR') |
       (df['street_suffix'] == 'TRL'), 'street_suffix'] = 'TRAIL'
df.loc[(df['street_suffix'] == 'TER') |
       (df['street_suffix'] == 'TERR'), 'street_suffix'] = 'TERRACE'
df.loc[(df['street_suffix'] == 'TPKE'), 'street_suffix'] = 'TURNPIKE'
df.loc[(df['street_suffix'] == 'BLVD'), 'street_suffix'] = 'BOULEVARD'
df.loc[(df['street_suffix'] == 'EXT') |
       (df['street_suffix'] == 'EXTN'), 'street_suffix'] = 'EXTENSION'
df.loc[(df['street_suffix'] == 'RDG') |
       (df['street_suffix'] == 'RDGE'), 'street_suffix'] = 'RIDGE'
df.loc[(df['street_suffix'] == 'HL'), 'street_suffix'] = 'HILL'
df.loc[(df['street_suffix'] == 'WY'), 'street_suffix'] = 'WAY'

In [48]:
#checking changes
df['street_suffix'].value_counts().loc[lambda x : x>1000]

ROAD         119483
STREET        74190
DRIVE         57574
AVENUE        44320
LANE          33778
COURT          8540
CIRCLE         6244
PLACE          5242
WAY            4448
TERRACE        4147
TRAIL          2889
VILLAGE        2846
BOULEVARD      1725
TURNPIKE       1500
EXTENSION      1419
RIDGE          1376
HILL           1186
Name: street_suffix, dtype: int64

In [49]:
#Turning all other suffixes into NaN for clarity of analysis
suffixes = ['ROAD',
            'STREET',
            'DRIVE',
            'AVENUE',
            'LANE',
            'COURT',
            'CIRCLE',
            'PLACE',
            'WAY',
            'TERRACE',
            'TRAIL',
            'VILLAGE',
            'BOULEVARD',
            'TURNPIKE',
            'EXTENSION',
            'RIDGE',
            'HILL']
df.loc[~df['street_suffix'].isin(suffixes), 'street_suffix'] = np.nan

In [50]:
#checking changes
df['street_suffix'].value_counts(dropna=False)

ROAD         119483
STREET        74190
NaN           64272
DRIVE         57574
AVENUE        44320
LANE          33778
COURT          8540
CIRCLE         6244
PLACE          5242
WAY            4448
TERRACE        4147
TRAIL          2889
VILLAGE        2846
BOULEVARD      1725
TURNPIKE       1500
EXTENSION      1419
RIDGE          1376
HILL           1186
Name: street_suffix, dtype: int64

In [51]:
#dropping address from the dataframe
df.drop(columns = ['address'], inplace = True)

In [52]:
#checking current data frame
df.head(20)

serial_number  list_year        town  assessed_value  sale_amount  \
0         2020348       2020     Ansonia          150500     325000.0   
1           20002       2020     Ashford          253000     430000.0   
2          200212       2020        Avon          130400     179900.0   
3          200243       2020        Avon          619290     890000.0   
4          200377       2020        Avon          862330    1447500.0   
5          200109       2020        Avon          847520    1250000.0   
7         2020313       2020      Berlin          412000     677500.0   
8          200097       2020     Bethany          511000     795000.0   
9           20139       2020      Bethel          171360     335000.0   
10         200086       2020   Bethlehem          168900     352000.0   
11        2000381       2020  Bloomfield          163730     250000.0   
12         200258       2020    Branford          530500     700000.0   
13         200804       2020    Branford          455100     650000.0   
14         201295       2020     Bristol          144340     249900.0   
15         200032       2020     Bristol           94920     155000.0   
16         200354       2020     Bristol          173740     299000.0   
17         200527       2020     Bristol           85680     183000.0   
18         201212       2020     Bristol           88060      91000.0   
19       20000050       2020  Brookfield          105110     160000.0   
20       20000179       2020  Brookfield           94770     170000.0   

    sales_ratio property_type residential_type  in_pandemic  month  year  \
0      0.463000    Commercial              NaN            1      9  2021   
1      0.588300   Residential    Single Family            1     10  2020   
2      0.724800   Residential            Condo            1      3  2021   
3      0.695800   Residential    Single Family            1      4  2021   
4      0.595700   Residential    Single Family            1      7  2021   
5      0.678000   Residential    Single Family            1     12  2020   
7      0.608100   Residential    Single Family            1      7  2021   
8      0.642700    Commercial              NaN            1      6  2021   
9      0.511500   Residential    Single Family            1     12  2020   
10     0.479800   Residential    Single Family            1      8  2021   
11     0.654900   Residential            Condo            1      9  2021   
12     0.757800   Residential    Single Family            1      1  2021   
13     0.700100    Commercial              NaN            1      9  2021   
14     0.577500   Residential    Single Family            1      9  2021   
15     0.612300   Residential    Single Family            1     10  2020   
16     0.581000   Residential    Single Family            1     12  2020   
17     0.468100   Residential       Two Family            1      3  2021   
18     0.967600   Residential    Single Family            1      8  2021   
19     0.656937   Residential            Condo            1     11  2020   
20     0.557400   Residential            Condo            1      2  2021   

    street_number street_suffix  
0           230.0        AVENUE  
1           390.0          ROAD  
2             5.0         DRIVE  
3           111.0         DRIVE  
4            70.0         DRIVE  
5            57.0         DRIVE  
7           216.0          ROAD  
8            23.0          ROAD  
9            16.0         DRIVE  
10           39.0          ROAD  
11            9.0         RIDGE  
12           34.0           NaN  
13          119.0        STREET  
14          609.0        STREET  
15            9.0         COURT  
16          391.0          LANE  
17           25.0        STREET  
18            8.0       TERRACE  
19            4.0         COURT  
20           72.0          LANE

In [53]:
#checking current shape
df.shape

(435179, 13)

>Notes: street_suffix was derrived from last split of address string. The most popular suffixes (more than 1,000 instances) were consolidated for variations according to USPS standards. Remaining entries with less than 1,000 instances were turned into NaN values for consistency of results.

## 4. Merging Data Sets

### Town data

In [54]:
#Left joining df_town to df
df = pd.merge(df, df_town, on='town', how='left', indicator=True)

In [55]:
#checking merge
df.head(20)

serial_number  list_year        town  assessed_value  sale_amount  \
0         2020348       2020     Ansonia          150500     325000.0   
1           20002       2020     Ashford          253000     430000.0   
2          200212       2020        Avon          130400     179900.0   
3          200243       2020        Avon          619290     890000.0   
4          200377       2020        Avon          862330    1447500.0   
5          200109       2020        Avon          847520    1250000.0   
6         2020313       2020      Berlin          412000     677500.0   
7          200097       2020     Bethany          511000     795000.0   
8           20139       2020      Bethel          171360     335000.0   
9          200086       2020   Bethlehem          168900     352000.0   
10        2000381       2020  Bloomfield          163730     250000.0   
11         200258       2020    Branford          530500     700000.0   
12         200804       2020    Branford          455100     650000.0   
13         201295       2020     Bristol          144340     249900.0   
14         200032       2020     Bristol           94920     155000.0   
15         200354       2020     Bristol          173740     299000.0   
16         200527       2020     Bristol           85680     183000.0   
17         201212       2020     Bristol           88060      91000.0   
18       20000050       2020  Brookfield          105110     160000.0   
19       20000179       2020  Brookfield           94770     170000.0   

    sales_ratio property_type residential_type  in_pandemic  month  year  \
0      0.463000    Commercial              NaN            1      9  2021   
1      0.588300   Residential    Single Family            1     10  2020   
2      0.724800   Residential            Condo            1      3  2021   
3      0.695800   Residential    Single Family            1      4  2021   
4      0.595700   Residential    Single Family            1      7  2021   
5      0.678000   Residential    Single Family            1     12  2020   
6      0.608100   Residential    Single Family            1      7  2021   
7      0.642700    Commercial              NaN            1      6  2021   
8      0.511500   Residential    Single Family            1     12  2020   
9      0.479800   Residential    Single Family            1      8  2021   
10     0.654900   Residential            Condo            1      9  2021   
11     0.757800   Residential    Single Family            1      1  2021   
12     0.700100    Commercial              NaN            1      9  2021   
13     0.577500   Residential    Single Family            1      9  2021   
14     0.612300   Residential    Single Family            1     10  2020   
15     0.581000   Residential    Single Family            1     12  2020   
16     0.468100   Residential       Two Family            1      3  2021   
17     0.967600   Residential    Single Family            1      8  2021   
18     0.656937   Residential            Condo            1     11  2020   
19     0.557400   Residential            Condo            1      2  2021   

    street_number street_suffix  population  latitude  longitude _merge  
0           230.0        AVENUE       18945   41.3462   -73.0790   both  
1           390.0          ROAD        4193   41.8731   -72.1216   both  
2             5.0         DRIVE       18821   41.8096   -72.8305   both  
3           111.0         DRIVE       18821   41.8096   -72.8305   both  
4            70.0         DRIVE       18821   41.8096   -72.8305   both  
5            57.0         DRIVE       18821   41.8096   -72.8305   both  
6           216.0          ROAD       20107   41.6215   -72.7457   both  
7            23.0          ROAD        5331   41.4218   -72.9970   both  
8            16.0         DRIVE       20287   41.3712   -73.4140   both  
9            39.0          ROAD        3408   41.6404   -73.2058   both  
10            9.0         RIDGE       21399   41.8265   -72

In [56]:
#checking merge
df.tail(20)

serial_number  list_year           town  assessed_value  sale_amount  \
435159         100012       2010     Woodbridge          284900     415000.0   
435160          10155       2010        Milford          269100     310000.0   
435161         100027       2010       Branford          111100     139000.0   
435162         100117       2010      Newington           65310      97000.0   
435163          10146       2010     Bloomfield          358470     560000.0   
435164          10212       2010         Hamden          181160     207500.0   
435165          10055       2010    North Haven          168980     190000.0   
435166         100846       2010       Stamford          738830     687500.0   
435167          10088       2010         Groton          114240     128000.0   
435168          10076       2010     Brookfield          349480     439000.0   
435169         100077       2010   Old Saybrook          318000     340000.0   
435170          10019       2010      Watertown          161600     170000.0   
435171         100170       2010  West Hartford          122700     187000.0   
435172         100358       2010      Stratford          160510     221000.0   
435173         100169       2010     Plainville          131740     149000.0   
435174         100125       2010        Norwich          305000     350000.0   
435175          10033       2010  South Windsor          242690     325000.0   
435176         100027       2010    East Granby           81900     113000.0   
435177         100266       2010     Farmington          537330     730000.0   
435178         100252       2010        Meriden          148070     188000.0   

        sales_ratio property_type residential_type  in_pandemic  month  year  \
435159     0.686506   Residential    Single Family            0     10  2010   
435160     0.868065   Residential            Condo            0     12  2010   
435161     0.799281   Residential            Condo            0     11  2010   
435162     0.673299   Residential            Condo            0     10  2010   
435163     0.640125   Residential    Single Family            0      6  2011   
435164     0.873060   Residential    Single Family            0      3  2011   
435165     0.889368   Residential    Single Family            0     12  2010   
435166     1.074662   Residential    Single Family            0      6  2011   
435167     0.892500   Residential    Single Family            0     12  2010   
435168     0.796082   Residential    Single Family            0      3  2011   
435169     0.935294   Residential    Single Family            0      3  2011   
435170     0.950588   Residential    Single Family            0     10  2010   
435171     0.656150   Residential    Single Family            0     12  2010   
435172     0.726290   Residential    Single Family            0      4  2011   
435173     0.884161   Residential            Condo            0      6  2011   
435174     0.871429   Residential    Single Family            0      1  2011   
435175     0.746738   Residential    Single Family            0     11  2010   
435176     0.724779   Residential            Condo            0     12  2010   
435177     0.736068   Residential    Single Family            0      7  2011   
435178     0.787606   Residential    Single Family            0      3  2011   

        street_number street_suffix  population  latitude  longitude _merge  
435159           34.0          ROAD        9073   41.3526   -73.0084   both  
435160            4.0          ROAD       52308   41.2307   -73.0640   both  
435161          346.0           NaN       28230   41.2799   -72.8141   both  
435162           85.0         DRIVE       30493   41.6973   -72.7228   both  
435163           19.0          LANE       21399   41.8265   -72.7301   both  
435164            4.0        CIRCLE       61160   41.3839   -72.9026   both  
435165            9.0          ROAD       24217   41.3903   -72.8595   both  
435166          143.0 

In [57]:
#checking merge counts
df['_merge'].value_counts()

both          435179
left_only          0
right_only         0
Name: _merge, dtype: int64

In [58]:
#dropping _merge flag
df.drop(columns = ['_merge'], inplace = True)

In [59]:
#checking current shape
df.shape

(435179, 16)

### Consumer Price Index

In [60]:
#creating key 'date' for merging CPI data
df.loc[df['month'] < 10, 'date'] = (df['year'].astype(str) + '-0' + df['month'].astype(str))
df.loc[df['month'] >= 10, 'date'] = (df['year'].astype(str) + '-' + df['month'].astype(str))

In [61]:
#checking key variable
df['date'].head(20)

0     2021-09
1     2020-10
2     2021-03
3     2021-04
4     2021-07
5     2020-12
6     2021-07
7     2021-06
8     2020-12
9     2021-08
10    2021-09
11    2021-01
12    2021-09
13    2021-09
14    2020-10
15    2020-12
16    2021-03
17    2021-08
18    2020-11
19    2021-02
Name: date, dtype: object

In [62]:
#Left joining df_cpi to df
df = pd.merge(df, df_cpi, on='date', how='left', indicator=True)

In [63]:
#checking merge
df.head(20)

serial_number  list_year        town  assessed_value  sale_amount  \
0         2020348       2020     Ansonia          150500     325000.0   
1           20002       2020     Ashford          253000     430000.0   
2          200212       2020        Avon          130400     179900.0   
3          200243       2020        Avon          619290     890000.0   
4          200377       2020        Avon          862330    1447500.0   
5          200109       2020        Avon          847520    1250000.0   
6         2020313       2020      Berlin          412000     677500.0   
7          200097       2020     Bethany          511000     795000.0   
8           20139       2020      Bethel          171360     335000.0   
9          200086       2020   Bethlehem          168900     352000.0   
10        2000381       2020  Bloomfield          163730     250000.0   
11         200258       2020    Branford          530500     700000.0   
12         200804       2020    Branford          455100     650000.0   
13         201295       2020     Bristol          144340     249900.0   
14         200032       2020     Bristol           94920     155000.0   
15         200354       2020     Bristol          173740     299000.0   
16         200527       2020     Bristol           85680     183000.0   
17         201212       2020     Bristol           88060      91000.0   
18       20000050       2020  Brookfield          105110     160000.0   
19       20000179       2020  Brookfield           94770     170000.0   

    sales_ratio property_type residential_type  in_pandemic  month  year  \
0      0.463000    Commercial              NaN            1      9  2021   
1      0.588300   Residential    Single Family            1     10  2020   
2      0.724800   Residential            Condo            1      3  2021   
3      0.695800   Residential    Single Family            1      4  2021   
4      0.595700   Residential    Single Family            1      7  2021   
5      0.678000   Residential    Single Family            1     12  2020   
6      0.608100   Residential    Single Family            1      7  2021   
7      0.642700    Commercial              NaN            1      6  2021   
8      0.511500   Residential    Single Family            1     12  2020   
9      0.479800   Residential    Single Family            1      8  2021   
10     0.654900   Residential            Condo            1      9  2021   
11     0.757800   Residential    Single Family            1      1  2021   
12     0.700100    Commercial              NaN            1      9  2021   
13     0.577500   Residential    Single Family            1      9  2021   
14     0.612300   Residential    Single Family            1     10  2020   
15     0.581000   Residential    Single Family            1     12  2020   
16     0.468100   Residential       Two Family            1      3  2021   
17     0.967600   Residential    Single Family            1      8  2021   
18     0.656937   Residential            Condo            1     11  2020   
19     0.557400   Residential            Condo            1      2  2021   

    street_number street_suffix  population  latitude  longitude     date  \
0           230.0        AVENUE       18945   41.3462   -73.0790  2021-09   
1           390.0          ROAD        4193   41.8731   -72.1216  2020-10   
2             5.0         DRIVE       18821   41.8096   -72.8305  2021-03   
3           111.0         DRIVE       18821   41.8096   -72.8305  2021-04   
4            70.0         DRIVE       18821   41.8096   -72.8305  2021-07   
5            57.0         DRIVE       18821   41.8096   -72.8305  2020-12   
6           216.0          ROAD       20107   41.6215   -72.7457  2021-07   
7            23.0          ROAD        5331   41.4218   -72.9970  2021-06   
8            16.0         DRIVE       20287   41.3712   -73.4140  2020-12   
9            39.0          ROAD        3408   41.6404   -73.2058  2021-08   
10            9.0         

In [64]:
#checking merge
df.tail(20)

serial_number  list_year           town  assessed_value  sale_amount  \
435159         100012       2010     Woodbridge          284900     415000.0   
435160          10155       2010        Milford          269100     310000.0   
435161         100027       2010       Branford          111100     139000.0   
435162         100117       2010      Newington           65310      97000.0   
435163          10146       2010     Bloomfield          358470     560000.0   
435164          10212       2010         Hamden          181160     207500.0   
435165          10055       2010    North Haven          168980     190000.0   
435166         100846       2010       Stamford          738830     687500.0   
435167          10088       2010         Groton          114240     128000.0   
435168          10076       2010     Brookfield          349480     439000.0   
435169         100077       2010   Old Saybrook          318000     340000.0   
435170          10019       2010      Watertown          161600     170000.0   
435171         100170       2010  West Hartford          122700     187000.0   
435172         100358       2010      Stratford          160510     221000.0   
435173         100169       2010     Plainville          131740     149000.0   
435174         100125       2010        Norwich          305000     350000.0   
435175          10033       2010  South Windsor          242690     325000.0   
435176         100027       2010    East Granby           81900     113000.0   
435177         100266       2010     Farmington          537330     730000.0   
435178         100252       2010        Meriden          148070     188000.0   

        sales_ratio property_type residential_type  in_pandemic  month  year  \
435159     0.686506   Residential    Single Family            0     10  2010   
435160     0.868065   Residential            Condo            0     12  2010   
435161     0.799281   Residential            Condo            0     11  2010   
435162     0.673299   Residential            Condo            0     10  2010   
435163     0.640125   Residential    Single Family            0      6  2011   
435164     0.873060   Residential    Single Family            0      3  2011   
435165     0.889368   Residential    Single Family            0     12  2010   
435166     1.074662   Residential    Single Family            0      6  2011   
435167     0.892500   Residential    Single Family            0     12  2010   
435168     0.796082   Residential    Single Family            0      3  2011   
435169     0.935294   Residential    Single Family            0      3  2011   
435170     0.950588   Residential    Single Family            0     10  2010   
435171     0.656150   Residential    Single Family            0     12  2010   
435172     0.726290   Residential    Single Family            0      4  2011   
435173     0.884161   Residential            Condo            0      6  2011   
435174     0.871429   Residential    Single Family            0      1  2011   
435175     0.746738   Residential    Single Family            0     11  2010   
435176     0.724779   Residential            Condo            0     12  2010   
435177     0.736068   Residential    Single Family            0      7  2011   
435178     0.787606   Residential    Single Family            0      3  2011   

        street_number street_suffix  population  latitude  longitude     date  \
435159           34.0          ROAD        9073   41.3526   -73.0084  2010-10   
435160            4.0          ROAD       52308   41.2307   -73.0640  2010-12   
435161          346.0           NaN       28230   41.2799   -72.8141  2010-11   
435162           85.0         DRIVE       30493   41.6973   -72.7228  2010-10   
435163           19.0          LANE       21399   41.8265   -72.7301  2011-06   
435164            4.0        CIRCLE       61160   41.3839   -72.9026  2011-03   
435165            9.0          ROAD       24217   41.3903   -72.8595  2010-12  

In [65]:
#checking merge counts
df['_merge'].value_counts()

both          435179
left_only          0
right_only         0
Name: _merge, dtype: int64

In [66]:
#dropping _merge flag
df.drop(columns = ['_merge', 'date'], inplace = True)

In [67]:
#checking drop results
df.head()

serial_number  list_year     town  assessed_value  sale_amount  \
0        2020348       2020  Ansonia          150500     325000.0   
1          20002       2020  Ashford          253000     430000.0   
2         200212       2020     Avon          130400     179900.0   
3         200243       2020     Avon          619290     890000.0   
4         200377       2020     Avon          862330    1447500.0   

   sales_ratio property_type residential_type  in_pandemic  month  year  \
0       0.4630    Commercial              NaN            1      9  2021   
1       0.5883   Residential    Single Family            1     10  2020   
2       0.7248   Residential            Condo            1      3  2021   
3       0.6958   Residential    Single Family            1      4  2021   
4       0.5957   Residential    Single Family            1      7  2021   

   street_number street_suffix  population  latitude  longitude  price_index  
0          230.0        AVENUE       18945   41.3462   -73.0790      269.086  
1          390.0          ROAD        4193   41.8731   -72.1216      254.076  
2            5.0         DRIVE       18821   41.8096   -72.8305      258.935  
3          111.0         DRIVE       18821   41.8096   -72.8305      261.237  
4           70.0         DRIVE       18821   41.8096   -72.8305      267.789

In [68]:
#checking current shape
df.shape

(435179, 17)

## 5. Inflation Normalizations
assessed_value and sale_amount normalizations will use the standard 1982=100 formula:<br>
**original value / price_index * 100 = normalized value**

### Normalized assessed_value

In [69]:
#creating variable for normalizd assessed_value rounded to 2 decimal places
df['norm_assessed_value'] = (df['assessed_value'] / df['price_index'] * 100).round(decimals=2)

In [70]:
#checking new variable stats.
print('Min:', df['norm_assessed_value'].min(),
      '   Max:', df['norm_assessed_value'].max(),
      '   Mean:', df['norm_assessed_value'].mean())

Min: 70.79    Max: 45231327.8    Mean: 104068.37271931741


### Normalized sale_amount

In [71]:
#creating variable for normalizd sale_amount rounded to 2 decimal places
df['norm_sale_amount'] = (df['sale_amount'] / df['price_index'] * 100).round(decimals=2)

In [72]:
#checking new variable stats.
print('Min:', df['norm_sale_amount'].min(),
      '   Max:', df['norm_sale_amount'].max(),
      '   Mean:', df['norm_sale_amount'].mean())

Min: 743.26    Max: 57507963.36    Mean: 156023.50849764099


### Normalized sales_ratio

In [73]:
#creating variable for normalizd sales_ratio
df['norm_sales_ratio'] = (df['norm_assessed_value'] / df['norm_sale_amount'])

In [74]:
#checking new variable stats.
print('Min:', df['norm_sales_ratio'].min(),
      '   Max:', df['norm_sales_ratio'].max(),
      '   Mean:', df['norm_sales_ratio'].mean())

Min: 0.046677045948789735    Max: 1.4084080232236653    Mean: 0.7094560737228637


In [75]:
#checking current shape
df.shape

(435179, 20)

In [76]:
#checking head and tail
df

serial_number  list_year           town  assessed_value  sale_amount  \
0             2020348       2020        Ansonia          150500     325000.0   
1               20002       2020        Ashford          253000     430000.0   
2              200212       2020           Avon          130400     179900.0   
3              200243       2020           Avon          619290     890000.0   
4              200377       2020           Avon          862330    1447500.0   
...               ...        ...            ...             ...          ...   
435174         100125       2010        Norwich          305000     350000.0   
435175          10033       2010  South Windsor          242690     325000.0   
435176         100027       2010    East Granby           81900     113000.0   
435177         100266       2010     Farmington          537330     730000.0   
435178         100252       2010        Meriden          148070     188000.0   

        sales_ratio property_type residential_type  in_pandemic  month  year  \
0          0.463000    Commercial              NaN            1      9  2021   
1          0.588300   Residential    Single Family            1     10  2020   
2          0.724800   Residential            Condo            1      3  2021   
3          0.695800   Residential    Single Family            1      4  2021   
4          0.595700   Residential    Single Family            1      7  2021   
...             ...           ...              ...          ...    ...   ...   
435174     0.871429   Residential    Single Family            0      1  2011   
435175     0.746738   Residential    Single Family            0     11  2010   
435176     0.724779   Residential            Condo            0     12  2010   
435177     0.736068   Residential    Single Family            0      7  2011   
435178     0.787606   Residential    Single Family            0      3  2011   

        street_number street_suffix  population  latitude  longitude  \
0               230.0        AVENUE       18945   41.3462   -73.0790   
1               390.0          ROAD        4193   41.8731   -72.1216   
2                 5.0         DRIVE       18821   41.8096   -72.8305   
3               111.0         DRIVE       18821   41.8096   -72.8305   
4                70.0         DRIVE       18821   41.8096   -72.8305   
...               ...           ...         ...       ...        ...   
435174          246.0          ROAD       40020   41.5243   -72.0759   
435175          142.0         DRIVE       26719   41.8490   -72.5718   
435176           16.0           NaN        5148   41.9412   -72.7273   
435177           45.0           NaN       26559   41.7360   -72.7950   
435178           40.0         DRIVE       60790   41.5382   -72.8070   

        price_index  norm_assessed_value  norm_sale_amount  norm_sales_ratio  
0           269.086             55930.07         120779.23          0.463077  
1           254.076             99576.50         169240.70          0.588372  
2           258.935             50360.13          69476.90          0.724847  
3           261.237            237060.60         340686.81          0.695831  
4           267.789            322018.45         540537.51          0.595737  
...             ...                  ...               ...               ...  
435174      216.400            140942.70         161737.52          0.871429  
435175      214.750            113010.48         151338.77          0.746738  
435176      215.262             38046.66          52494.17          0.724779  
435177      222.686            241294.92         327815.85          0.736068  
435178      220.024             67297.20          85445.22          0.787606  

[435179 rows x 20 columns]

# 6. Export

In [77]:
#exporting wrangled dataset
df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'CTRE 2011-2021 WRANGLED.pkl'))